File: credit.c
//Takes a valid number greater than 0 and determines the type of card if it is a valid credit number

#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

string checkSum(string type, int digits[], int length);

int main(void)
{
    long long creditNum;
    do
    {
        //gets a number greater than 0
        creditNum = get_long_long("Number: ");
    }
    while (creditNum <= 0);

    //declaration of necessary variables, also finds the length of the cc number
    int length = floor(log10(creditNum)) + 1;
    int digits[length];
    string type = "";

    for (int i = 0; i < length; i ++)
    {
        //adds each digit seperately to an array
        digits[length - i - 1] = creditNum % 10;
        creditNum = creditNum / 10;
    }

    if (length == 15 && digits[0] == 3 && (digits[1] == 4 || digits[1] == 7))
    {
        //15 digits, begins with 34 or 37: is an AMEX unless checksum fails
        type = "AMEX";
    }
    else if ((length == 13 || length == 16) && digits[0] == 4)
    {
        //13 or 16 digits, begins with 4: is a VISA unless checksum fails
        type = "VISA";
    }
    else if (length == 16 && digits[0] == 5 && (digits[1] >= 1 && digits[1] <= 5))
    {
        //16 digits and begins with 5 and a second digit between 1 and 5: MASTERCARD
        type = "MASTERCARD";
    }
    else
    {
        //otherwise, invalid
        type = "INVALID";
    }

    if (strcmp(type, "INVALID") != 0)
    {
        type = checkSum(type, digits, length);
    }

    printf("%s \n", type);
}

string checkSum(string type, int digits[], int length)
{
    int counter = 0;
    int total = 0;
    for (int i = length - 1; i >= 0; i --)
    {
        int product;
        if (counter % 2 != 0)
        {
            product = 2 * digits[i];
            int tendigit = (product - product % 10);
            total = total + tendigit / 10 + product - tendigit;
            // printf("%i : %i \n", digits[i], total);
        }
        else
        {
            total = total + digits[i];
        }
        counter ++;
    }

    if (total % 10 != 0)
    {
        return "INVALID";
    }
    else
    {
        return type;
    }
}

File: credit.py
# credit.py
# Determine whether a credit card number is valid

from cs50 import get_int

# Constants related to each type of card
AMEX_1 = 34
AMEX_2 = 37
AMEX_LENGTH = 15

FIRST_MC = 51
LAST_MC = 55
MC_LENGTH = 16

FIRST_VISA = 40
LAST_VISA = 49
VISA_LENGTH1 = 13
VISA_LENGTH2 = 16


def main():
    # Request number
    cc_no = get_int('Number: ')

    checksum = 0
    first_two_digits = 0
    mult_by_two = False
    cc_no_length = 0

    while cc_no > 0:
        cur_digit = cc_no % 10

        # Add new digit to checksum, double if necessary
        if mult_by_two:
            if cur_digit < 5:
                # Normal case: don't have to worry about digits
                checksum += cur_digit * 2
            else:
                # Number is >= 5, so we must subtract 9 to get the sum of the digits
                checksum += (cur_digit * 2 - 9)

        else:
            checksum += cur_digit

        # Keep track of first two digits
        first_two_digits = cur_digit * 10 + (first_two_digits // 10)

        # Update state and drop last digit
        mult_by_two = not mult_by_two
        cc_no = cc_no // 10
        cc_no_length += 1

    # Validade checksum, first two digits, and the length
    if checksum % 10 != 0:
        print('INVALID')
    elif cc_no_length == AMEX_LENGTH and (first_two_digits == AMEX_1 or first_two_digits == AMEX_2):
        print('AMEX')
    elif first_two_digits >= FIRST_VISA and first_two_digits <= LAST_VISA and (cc_no_length == VISA_LENGTH1 or cc_no_length == VISA_LENGTH2):
        print('VISA')
    elif first_two_digits >= FIRST_MC and first_two_digits <= LAST_MC and cc_no_length == MC_LENGTH:
        print('MASTERCARD')
    else:
        print('INVALID')


if __name__ == "__main__":
    main()
